core/domain/expression_parser.py

```diff
@@ -0,0 +1,509 @@
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A parser that generates a parse tree for math expressions.
+
+It uses the following grammar in Backus-Naur form:
+
+<expr> ::= <expr2> (('+' | '-') <expr2>)*
+<expr2> ::= <expr1> (('*' | '/') <expr1>)*
+<expr1> ::= '-' <expr1> | '+' <expr1> | <unit> ('^' <expr1>)?
+
+<unit> ::= <identifier> | <number> | '(' <expr> ')' | <function> '(' <expr> ')'
```

Comment:
Is "2x" a valid unit/expression? It seems missing from your tests, so I cannot tell.

Reply:
No, it is not. I've added tests for it now.
Though note that, I'd added a test with 'ab/2' which seemed to cover this case.

Reply:
I think I saw that, but note that "ab" is different from "2x" conceptually.